lib/puppet/functions/any_resources_of_type.rb

```diff
@@ -0,0 +1,25 @@
+Puppet::Functions.create_function(:any_resources_of_type, Puppet::Functions::InternalFunction) do 
+  dispatch :any_resources_of_type do
+    scope_param
+
+    required_param 'String', :resource_type
+    optional_param 'Hash[Any, Any]', :parameters
+  end
+
+  def any_resources_of_type(scope, resource_type, parameters = nil)
+    scope.catalog.resources.any? do |resource|
+      # We should always iterate over Puppet::Parser::Resource
+      # instances here, and documentation states that types can be
+      # strings or symbols.
+      # https://www.rubydoc.info/gems/puppet/Puppet/Resource#initialize-instance_method
+      if resource.type.to_s.casecmp(resource_type).zero? # String#casecmp? is Ruby 2.4+
```

Comment:
I had a concern about calling `to_s` and increase memory usage if there are a lot of resources,  but i think Ruby is smart enough and if you call `to_s` on a String, it will not create a new object as it does when calling it on other Types.